common/src/main/java/com/wynntils/screens/skillpointloadouts/SkillPointLoadoutsScreen.java

```diff
@@ -509,6 +450,47 @@ public boolean doMouseClicked(double mouseX, double mouseY, int button) {
         return super.doMouseClicked(mouseX, mouseY, button);
     }
 
+    // baseYPosition - dividedHeight * maxScrollOffset * scrollPercent = dividedHeight * 52
+    // Solve for maxScrollOffset
+    // Full explanation in #artemis-dev
+    @Override
+    public boolean mouseScrolled(double mouseX, double mouseY, double scrollX, double scrollY) {
+        // 11 loadouts are fully displayed from 9 to 56
+        // 12th one is very slightly cut off (it needs 57)
+        if (loadoutWidgets.size() <= MAX_LOADOUTS_PER_PAGE) {
+            return super.mouseScrolled(mouseX, mouseY, scrollX, scrollY);
+        }
+
+        int scrollableWidgets = Math.max(0, loadoutWidgets.size() - MAX_LOADOUTS_PER_PAGE);
+        float scrollableRatio = (float) scrollableWidgets / loadoutWidgets.size();
+        float maxScrollOffset = (4 * (loadoutWidgets.size() - 1) - 43) / scrollableRatio;
+        scrollPercent = (float) Math.max(0, Math.min(scrollableRatio, scrollPercent - scrollY / 50));
+
+        loadoutWidgets.forEach(widget -> {
+            float baseYPosition = dividedHeight * (9f + loadoutWidgets.indexOf(widget) * 4f);
+
+            float scrollOffset = dividedHeight * maxScrollOffset * scrollPercent;
+            widget.setY((int) (baseYPosition - scrollOffset));
+            widget.visible = !(widget.getY() <= dividedHeight * 4) && !(widget.getY() >= dividedHeight * 56);
+        });
+
+        return super.mouseScrolled(mouseX, mouseY, scrollX, scrollY);
+    }
+
+    @Override
+    public boolean keyPressed(int keyCode, int scanCode, int modifiers) {
+        if (keyCode == GLFW.GLFW_KEY_DELETE && deleteButton.active) {
+            deleteButton.onPress();
+            return true;
+        } else if (keyCode == GLFW.GLFW_KEY_END) {
+            mouseScrolled(0, 0, 0, Float.NEGATIVE_INFINITY);
```

Comment:
I can't say I like doing this call. Maybe separate the scroll logic to a separate method from mouseScrolled, then call that here?